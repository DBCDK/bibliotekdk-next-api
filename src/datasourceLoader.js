import DataLoader from "dataloader";
import { log } from "dbc-node-logger";
import { withRedis } from "./datasources/redis.datasource";
import monitor from "./utils/monitor";
import { getFilesRecursive } from "./utils/utils";

// Find all datasources in src/datasources
export const datasources = getFilesRecursive("./src/datasources")
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status } = require(file.path);
    if (!load) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    // Monitor the load function from the datasource
    const monitoredLoad = monitor(
      {
        name: `REQUEST_${name}`,
        help: `${name} requests`,
      },
      load
    );

    // if datasource exports a createBatchLoader we use that,
    // otherwise we use default batch loader
    let monitoredBatchLoader = batchLoader
      ? (keys) => batchLoader(keys, monitoredLoad)
      : async (keys) => {
          return await Promise.all(keys.map((key) => monitoredLoad(key)));
        };

    // Check if Redis is configured for this datasource
    if (options && options.redis && options.redis.prefix && options.redis.ttl) {
      monitoredBatchLoader = withRedis(monitoredBatchLoader, {
        prefix: options.redis.prefix,
        ttl: options.redis.ttl,
      });
    }

    const statusChecker = status && (async () => await status(monitoredLoad));

    return {
      batchLoader: monitoredBatchLoader,
      name,
      options,
      statusChecker,
    };
  })
  .filter((func) => !!func);

log.info(`found ${datasources.length} datasources`, {
  datasources: datasources.map((datasource) => ({
    name: datasource.name,
    options: datasource.options,
  })),
});

/**
 * Will instantiate dataloaders from datasources
 */
export default function createDataLoaders() {
  const result = {};
  datasources.forEach((datasource) => {
    result[datasource.name] = new DataLoader(datasource.batchLoader, {
      // If key is an object, we stringify
      // to make it useful as a cache key
      cacheKeyFn: (key) =>
        typeof key === "object" ? JSON.stringify(key) : key,
    });
  });
  return result;
}
