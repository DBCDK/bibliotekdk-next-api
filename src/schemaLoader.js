/**
 * @file Responsible for loading the schema
 *
 * It will check all files recursively in ./src/schema
 * and load type definitions and resolvers
 */

import { makeExecutableSchema, mergeSchemas } from "graphql-tools";
import drupalSchema from "./schema/external/drupal";
import { log } from "dbc-node-logger";
import { getFilesRecursive } from "./utils/utils";

/**
 * Will check all files in schema folder
 * for type definitions and resolvers.
 *
 * These are then loaded
 */
function schemaLoader() {
  let allTypeDefs = [];
  let allResolvers = {};

  // Load files in schema folder
  const files = getFilesRecursive("./src/schema");

  // Require typeDefs and resolvers
  files.forEach((file) => {
    if (!file.path.endsWith(".js")) {
      return;
    }
    const { typeDef, resolvers } = require(file.path);
    if (typeDef) {
      allTypeDefs = [...allTypeDefs, typeDef];
      log.info(`Found type definition in ${file.path}`);
    }
    if (resolvers) {
      allResolvers = { ...allResolvers, ...resolvers };
    }
  });

  return { typeDefs: allTypeDefs, resolvers: allResolvers };
}

const { typeDefs, resolvers } = schemaLoader();

/**
 * Create executable schema from type definitions and resolvers
 */
export const internalSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

/**
 * We stitch the internal schema together with the external drupal schema
 */
export default async () => {
  return mergeSchemas({
    subschemas: [{ schema: await drupalSchema() }, { schema: internalSchema }],
  });
};
