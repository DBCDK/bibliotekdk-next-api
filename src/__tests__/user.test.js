import mockedLibrary from "../datasources/mocked/library.datasource.mocked";
import mockedUser from "../datasources/mocked/user.datasource.mocked";

import { graphql } from "graphql";
import { internalSchema } from "../schemaLoader";
import { createMockedDataLoaders } from "../datasourceLoader";

export async function performTestQuery({ query, variables, context }) {
  return graphql(internalSchema, query, null, context, variables);
}

test("user - get basic data", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                name
                address
                postalCode
                mail
                culrMail
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});

test("user - get agency danish", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                name
                address
                postalCode
                mail
                agency(language: da) {
                    branches {
                      openingHours
                      name
                    }
                  }
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});

test("user - get agency english", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                name
                address
                postalCode
                mail
                agency(language: en) {
                    branches {
                      openingHours
                      name
                    }
                  }
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});

test("user - get loans", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                loans {
                  loanId
                  dueDate   
                }
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});

test("user - get orders", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                orders {
                  orderId
                  status
                  pickUpExpiryDate
                  orderDate
                }
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});

test("user - get debt", async () => {
  const result = await performTestQuery({
    query: `
            query  {
              user {
                debt {
                  amount
                  creator
                  currency
                  date
                  title
                }
              }
            }
        `,
    variables: {},
    context: {
      datasources: createMockedDataLoaders(),
      accessToken: "DUMMY_TOKEN",
    },
  });
  expect(result).toMatchSnapshot();
});
