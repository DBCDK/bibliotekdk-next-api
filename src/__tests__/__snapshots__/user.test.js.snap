// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user - get agency danish 1`] = `
Object {
  "data": Object {
    "user": Object {
      "address": "Borgmesterskoven 45",
      "agency": Object {
        "branches": Array [
          Object {
            "name": "Hovedbiblioteket, Krystalgade",
            "openingHours": "Åbningstider
Se åbningstider på:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Blågårdens Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Brønshøj Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Christianshavns Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Husum Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Islands Brygge Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Øbro Jagtvej Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Biblioteket Rentemestervej",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Nørrebro Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Solvang Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Sundby Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Bibliotekshuset",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Sydhavnens Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Tingbjerg Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Valby Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Kulturstationen Vanløse. Biblioteket",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Vesterbro Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Vigerslev Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Østerbro Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Ørestad Bibliotek",
            "openingHours": "Åbningstider
Se åbningstider på:
bibliotek.kk.dk/biblioteker/",
          },
          Object {
            "name": "Københavns Bibliotekers centrale sortering",
            "openingHours": "ma-fre: 9-16",
          },
        ],
      },
      "mail": null,
      "name": "Freja Damgaard",
      "postalCode": "8660",
    },
  },
}
`;

exports[`user - get agency english 1`] = `
Object {
  "data": Object {
    "user": Object {
      "address": "Borgmesterskoven 45",
      "agency": Object {
        "branches": Array [
          Object {
            "name": "Copenhagen City Public Library",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Blågård",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Brønshøj",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Christianshavn",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Husum",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Islands Brygge",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Jagtvej",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Rentemestervej",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Nørrebro",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Solvang Bibliotek",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Sundby",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Bibliotekshuset",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Sydhavnen",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Tingbjerg",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Valby",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Vanløse",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Vesterbro",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Vigerslev",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Østerbro",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Ørestad Bibliotek",
            "openingHours": "Opening hours
See opening hours on:
<a href=\\"https://bibliotek.kk.dk/biblioteker/\\">bibliotek.kk.dk/biblioteker</a>",
          },
          Object {
            "name": "Københavns Bibliotekers centrale sortering",
            "openingHours": "Monday-Friday from 9 am to 4 pm. ",
          },
        ],
      },
      "mail": null,
      "name": "Freja Damgaard",
      "postalCode": "8660",
    },
  },
}
`;

exports[`user - get basic data 1`] = `
Object {
  "data": Object {
    "user": Object {
      "address": "Borgmesterskoven 45",
      "culrMail": "some@mail.com",
      "mail": null,
      "name": "Freja Damgaard",
      "postalCode": "8660",
    },
  },
}
`;

exports[`user - get debt 1`] = `
Object {
  "data": Object {
    "user": Object {
      "debt": Array [
        Object {
          "amount": "224",
          "creator": null,
          "currency": "DKK",
          "date": 1969-12-31T23:00:00.000Z,
          "title": "",
        },
        Object {
          "amount": "50",
          "creator": null,
          "currency": "DKK",
          "date": 1969-12-31T23:00:00.000Z,
          "title": "",
        },
      ],
    },
  },
}
`;

exports[`user - get loans 1`] = `
Object {
  "data": Object {
    "user": Object {
      "loans": Array [
        Object {
          "dueDate": 2021-05-01T22:00:00.000Z,
          "loanId": "120200589",
        },
        Object {
          "dueDate": 2021-04-29T22:00:00.000Z,
          "loanId": "120200590",
        },
      ],
    },
  },
}
`;

exports[`user - get orders 1`] = `
Object {
  "data": Object {
    "user": Object {
      "orders": Array [
        Object {
          "orderDate": 2021-04-21T11:37:49.000Z,
          "orderId": "12341234",
          "pickUpExpiryDate": 2021-05-02T22:00:00.000Z,
          "status": "AT_RESERVATION_SHELF",
        },
      ],
    },
  },
}
`;
